sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java

```diff
@@ -1526,15 +1748,28 @@ static class DatastoreWriterFn extends DoFn<Mutation, Void> {
         @Nullable String localhost,
         V1DatastoreFactory datastoreFactory,
         WriteBatcher writeBatcher) {
+      this(projectId, null, localhost, datastoreFactory, writeBatcher);
+    }
+
+    @VisibleForTesting
+    DatastoreWriterFn(
+        ValueProvider<String> projectId,
+        ValueProvider<String> databaseIdProvider,
+        @Nullable String localhost,
+        V1DatastoreFactory datastoreFactory,
+        WriteBatcher writeBatcher) {
       this.projectId = checkNotNull(projectId, "projectId");
+      this.databaseId = databaseIdProvider == null ? DEFAULT_DATABASE : databaseIdProvider.get();
```

Comment:
`.get()` here defeats the purpose of ValueProviders. Even if it's null, it doesn't mean that it is accessible (e.g., in case of RuntimeValueProvider).

Ideally you would `.get()` only inside a DoFn. My suggestion would be to keep it using the same thing as we do for `projectId`.